README.md
Python interface to ACM algorithm 772 (STRIPACK) fortran code for triangulation
on a sphere.  Includes routines from algorithm 623 for interpolation
(nearest neighbor, linear and hermite cubic).

to install ([numpy](http://numpy.org) and fortran compiler (preferably 
[gfortran](https://gcc.gnu.org/wiki/GFortran)) required):

* ``python setup.py build``
   - If you change the fortran compiler, you may have to add the 
flags `config_fc --fcompiler=<compiler name>` when setup.py is run
(see docs for [numpy.distutils] (http://docs.scipy.org/doc/numpy-dev/f2py/distutils.html)).
* ``python setup.py install``

* to run test, execute ``python stripack/__init__.py``

* see source code [docstrings](https://github.com/jswhit/stripack/blob/master/stripack/__init__.py)
for documentation,
[``test/fib_test.py``](https://github.com/jswhit/stripack/blob/master/test/fib_test.py) for example usage.

* see sample codes in ``fv3`` and ``mom6`` directories for applications to FV3 and MOM6 grids.

References:

 R. J. Renka, "ALGORITHM 623:  Interpolation on the Surface of a
 Sphere", ACM Trans. Math. Software, Vol. 10, No. 4, December 1984,
 pp. 437-439. (http://dl.acm.org/citation.cfm?id=356107)

 R. J. Renka, "ALGORITHM 772: STRIPACK: Delaunay triangulation
 and Voronoi diagram on the surface of a sphere"
 ACM Trans. Math. Software, Volume 23 Issue 3, Sept. 1997
 pp 416-434. (http://dl.acm.org/citation.cfm?id=275329)


Changelog
10/2/2021: version 1.2
-----------------------
* added 'fv3' directory, removed old codes in test dir.
* replaced travis with github workflow

12/22/2015: version 1.1 
-----------------------
* added support for hermite cubic interpolation.

12/20/2015: version 1.0 
-----------------------
* support for triangulation, linear and nearest neighbor interpolation.


test/fib_test.py
try:
    import matplotlib.pyplot as plt
    from mpl_toolkits.basemap import Basemap
    has_matplotlib = True
except:
    has_matplotlib = False
import numpy as np
from stripack import trmesh
import time

def fibonacci_pts(npts):
    # return lats and lons of N=npts fibonacci grid on a sphere.
    pi = np.pi
    inc = pi * (3.0 - np.sqrt(5.0))
    off = 2. / npts
    lats = []; lons = []
    for k in range(npts):
       y = k*off - 1. + 0.5*off
       r = np.sqrt(1 - y**2)
       phi = k * inc
       x = np.cos(phi)*r
       z = np.sin(phi)*r
       theta = np.arctan2(np.sqrt(x**2+y**2),z)
       phi = np.arctan2(y,x)
       lats.append( 0.5*pi-theta )
       if phi < 0.: phi = 2.*pi+phi
       lons.append( phi )
    return np.array(lats), np.array(lons)


# fake test data.
def test_func(lon, lat):
    nexp = 8
    return np.cos(nexp*lon)*np.sin(0.5*lon)**nexp*np.cos(lat)**nexp+np.sin(lat)**nexp
npts = 100000
lats, lons = fibonacci_pts(npts)
shuffle = True # shuffling the points speeds up the triangulation
if shuffle:
    ix = np.arange(npts)
    np.random.shuffle(ix)
    lats = lats[ix]; lons = lons[ix]
icos_data = test_func(lons,lats)

t1 = time.time()
print('triangulation of', len(lons),' points')
tri = trmesh(lons, lats)
print('triangulation took',time.time()-t1,' secs')

nlons = 1440; nlats = nlons/2 + 1 # 1 degree output mesh
delta = 360./nlons
olons = delta*np.arange(nlons)
olats = -90.0 + delta*np.arange(nlats)
olons = np.radians(olons);  olats = np.radians(olats)
olons, olats = np.meshgrid(olons, olats)

t1 = time.time()
order = 1 # can be 0 (nearest neighbor) or 1 (linear)
latlon_data = tri.interp(olons,olats,icos_data,order=order)
print('interpolation took',time.time()-t1,' secs')

latlon_datax = test_func(olons,olats)
print('max abs error:',(np.abs(latlon_datax-latlon_data)).max())
print('min/max field:',latlon_data.min(), latlon_datax.max())

# make plot on output mesh
if has_matplotlib:
    fig = plt.figure(figsize=(12,6))
    fig.add_subplot(1,2,1)
    m = Basemap(projection='ortho',lon_0=180,lat_0=20)
    x,y = m(np.degrees(olons), np.degrees(olats))
    m.drawcoastlines()
    m.drawmapboundary()
    m.contourf(x,y,latlon_data,15)
    plt.title('interpolated field order=%s' % order)
    m.colorbar()
    fig.add_subplot(1,2,2)
    m.drawcoastlines()
    m.drawmapboundary()
    m.contourf(x,y,latlon_data-latlon_datax,15)
    plt.title('error')
    m.colorbar()
    plt.tight_layout()
    plt.savefig('fib_test.png')
    plt.show()


test/test_fv3.py
from __future__ import print_function
from netCDF4 import Dataset
import numpy as np
import time
from stripack import trmesh
try:
    import cPickle
except:
    import _pickle as cPickle

# test fv3 interpolation from native history files to random points.

res = 128
fixfv3 = '/scratch4/NCEPDEV/global/save/glopara/svn/fv3gfs/fix/fix_fv3_gmted2010'
#fixfv3='/lustre/f1/unswept/Jeffrey.S.Whitaker/fv3_reanl/fv3gfs/global_shared.v15.0.0/fix/fix_fv3_gmted2010'
datapath = '/scratch3/BMC/gsienkf/whitaker/fv3_ics64'
#datapath = '/lustre/f1/unswept/Jeffrey.S.Whitaker/fv3_reanl/ics'
date = '2016100100'
var = 'pressfc'
ntime = 0


# perform triangulation.
lons = []; lats = []
for ntile in range(1,7,1):
    gridfile = '%s/C%s/C%s_grid.tile%s.nc'% (fixfv3,res,res,ntile)
    nc = Dataset(gridfile)
    lonsmid = nc['x'][1::2,1::2]
    latsmid = nc['y'][1::2,1::2]
    lons.append(lonsmid); lats.append(latsmid)
    nc.close()
lons = np.radians(np.array(lons,dtype=np.float64)).ravel()
lats = np.radians(np.array(lats,dtype=np.float64)).ravel()
t1 = time.clock()
print('triangulation of', len(lons),' points')
tri = trmesh(lons, lats)
print('triangulation took',time.clock()-t1,' secs')

# pickle it.
picklefile = 'C%s_grid.pickle' % res
cPickle.dump(tri,open(picklefile,'wb'),-1)
# load from pickle
t1 = time.clock()
tri2 = cPickle.load(open(picklefile,'rb'))
print('load from pickle file took',time.clock()-t1,' secs')

# generate random points on a sphere
# so that every small area on the sphere is expected
# to have the same number of points.
# http://mathworld.wolfram.com/SpherePointPicking.html
npts = 10000
np.random.seed(42) # set random seed for reproducibility
u = np.random.uniform(0.,1.,size=npts)
v = np.random.uniform(0.,1.,size=npts)
olons = np.radians(360.*u)
olats = np.radians((180./np.pi)*np.arccos(2*v-1) - 90.)

# read data from history files.
t1 = time.clock()
data = []
for ntile in range(1,7,1):
    datafile = '%s/C%s_%s/mem001/fv3_history2d.tile%s.nc'% (datapath,res,date,ntile)
    nc = Dataset(datafile)
    data.append(nc[var][ntime,...])
    nc.close()
data = (np.array(data,dtype=np.float64)).ravel()
print('read %s data from history files took' % var,time.clock()-t1,' secs')

# interpolate to random points.
t1 = time.clock()
data_interp = tri2.interp_linear(olons,olats,data)
print('interpolation %s points took' % npts,time.clock()-t1,' secs')
print('min/max interpolated field:',data_interp.min(), data_interp.max())


fv3/test_interp.py
from netCDF4 import Dataset
import numpy as np
from stripack import trmesh
import sys, time
import _pickle as cPickle

if len(sys.argv) > 1:
    res = int(sys.argv[1])
else:
    raise SystemExit("input cubed sphere resolution")

# path to FV3 fix files (to read orography and test interpolation)
fixfv3 = '/scratch1/NCEPDEV/global/glopara/fix_NEW/fix_fv3_gmted2010/'

# perform interpolation of orography using saved triangulation.

# first read data from each of the 6 tile files, append into 1d array
data = []
for ntile in range(1,7,1):
    gridfile = '%s/C%s/C%s_oro_data.tile%s.nc'% (fixfv3,res,res,ntile)
    nc = Dataset(gridfile)
    data.append(nc['orog_filt'][:,:])
    nc.close()
data = (np.array(data,dtype=np.float64)).ravel()

# then load trmesh object from pickle (pre-generated using generate_fv3mesh.py)

picklefile = 'C%s_grid.pickle' % res
tri = cPickle.load(open(picklefile,'rb'))

# do interpolation.
lons2d,lats2d,latlon_data = tri.interp_latlon(1440,data) # 1/4 deg output grid

# make a plot.
import cartopy.crs as ccrs
import matplotlib.pyplot as plt
fig = plt.figure()
ax = plt.axes(projection=ccrs.PlateCarree(central_longitude=0))
ax.contourf(lons2d,lats2d,latlon_data,15)
plt.savefig('test_interp.png')


fv3/generate_fv3mesh.py
from netCDF4 import Dataset
import numpy as np
import sys
from stripack import trmesh
import _pickle as cPickle

# generate FV3 mesh triangulation objects, save to pickle files.

if len(sys.argv) > 1:
    res = int(sys.argv[1])
else:
    raise SystemExit("input cubed sphere resolution")

shuffle = True
if len(sys.argv) > 2:
    shuffle = bool(int(sys.argv[1]))
print('generating triangulation for C%s' % res)
if shuffle: print('grid points randomly shuffled')

# path to FV3 fix files 
fixfv3 = '/scratch1/NCEPDEV/global/glopara/fix_NEW/fix_fv3_gmted2010/'

# read in lat/lon of native tile points from oro_data files.
lons = []; lats = []
for ntile in range(1,7,1):
    gridfile = '%s/C%s/C%s_oro_data.tile%s.nc'% (fixfv3,res,res,ntile)
    nc = Dataset(gridfile)
    lons2 = nc['geolon'][:,:]
    lats2 = nc['geolat'][:,:]
    lons.append(lons2); lats.append(lats2)
    nc.close()
lons = np.radians(np.array(lons,dtype=np.float64)).ravel()
lats = np.radians(np.array(lats,dtype=np.float64)).ravel()

# randomly shuffle points (may speed up triangulation)
ix = np.arange(len(lons))
if shuffle:
    np.random.shuffle(ix)
    lons = lons[ix]; lats = lats[ix]

# generate mesh triangulation object
tri = trmesh(lons, lats)
tri._shuffle = shuffle
tri._ix = ix

# pickle it and save it.
picklefile = 'C%s_grid.pickle' % res
cPickle.dump(tri,open(picklefile,'wb'),-1)


mom6/generate_mom6mesh.py
from netCDF4 import Dataset
import numpy as np
import sys
from stripack import trmesh
import _pickle as cPickle

# generate mom6 mesh triangulation objects, save to pickle files.

if len(sys.argv) > 1:
    res = sys.argv[1]
else:
    raise SystemExit("input mom6 resolution (025 for 0.25 deg)")

shuffle = True
if len(sys.argv) > 2:
    shuffle = bool(int(sys.argv[1]))
print('generating triangulation for mx%s' % res)
if shuffle: print('grid points randomly shuffled')

# path to MOM6 geometry file 
mom6geom = './ocean_geometry_mx%s.nc' % res

# read in lat/lon of native tile points from oro_data files.
nc = Dataset(mom6geom)
lons = np.radians(nc['geolon'][:,:].astype(np.float64)).ravel()
lats = np.radians(nc['geolat'][:,:].astype(np.float64)).ravel()
nc.close()

# randomly shuffle points (may speed up triangulation)
ix = np.arange(len(lons))
if shuffle:
    np.random.shuffle(ix)
    lons = lons[ix]; lats = lats[ix]

# generate mesh triangulation object
tri = trmesh(lons, lats)
tri._shuffle = shuffle
tri._ix = ix

# pickle it and save it.
picklefile = 'mom6_mx%s_grid.pickle' % res
cPickle.dump(tri,open(picklefile,'wb'),-1)


stripack/__init__.py
import  _stripack
import time
import numpy as np

__version__ = "1.2"

class trmesh(object):
    def __init__(self, lons, lats):
        """
given mesh points (lons,lats in radians) define triangulation.
n is size of input mesh (length of 1-d arrays lons and lats).

Algorithm:
 R. J. Renka, "ALGORITHM 772: STRIPACK: Delaunay triangulation
 and Voronoi diagram on the surface of a sphere"
 ACM Trans. Math. Software, Volume 23 Issue 3, Sept. 1997
 pp 416-434.

Instance variables:

lats, lons: lat/lon values of N nodes (in radians)
npts:  number of nodes (N).
x, y, z: cartesian coordinates corresponding to lats, lons.

lst: 6*(N-2) nodal indices, which along with
with lptr and lend, define the triangulation as a set of N
adjacency lists; counterclockwise-ordered sequences of neighboring nodes
such that the first and last neighbors of a boundary node are boundary
nodes (the first neighbor of an interior node is arbitrary).  In order to
distinguish between interior and boundary nodes, the last neighbor of
each boundary node is represented by the negative of its index.
The indices are 1-based (as in Fortran), not zero based (as in python).

lptr:  Set of 6*(N-2) pointers (indices)
in one-to-one correspondence with the elements of lst.
lst(lptr(i)) indexes the node which follows lst(i) in cyclical
counterclockwise order (the first neighbor follows the last neighbor).
The indices are 1-based (as in Fortran), not zero based (as in python).

lend: N pointers to adjacency lists.
lend(k) points to the last neighbor of node K.  lst(lend(K)) < 0 if and
only if K is a boundary node.
The indices are 1-based (as in Fortran), not zero based (as in python).
 """
        if len(lons.shape) != 1 or len(lats.shape) != 1:
            raise ValueError('lons and lats must be 1d')
        lats = lats.astype(np.float64,copy=False)
        lons = lons.astype(np.float64,copy=False)
        lons = lons.clip(-2.*np.pi,2.*np.pi)
        if (np.abs(lons)).max() > 2.*np.pi:
            msg="lons must be in radians (-2*pi <= lon <= 2*pi)"
            raise ValueError(msg)
        if (np.abs(lats)).max() > 0.5*np.pi:
            msg="lats must be in radians (-pi/2 <= lat <= pi/2)"
            raise ValueError(msg)
        npts = len(lons)
        if len(lats) != npts:
            raise ValueError('lons and lats must have same length')
        # compute cartesian coords on unit sphere.
        x,y,z = _stripack.trans(lats,lons,npts)
        lst,lptr,lend,ierr = _stripack.trmesh(x,y,z,npts)
        if ierr != 0:
            raise ValueError('ierr = %s in trmesh' % ierr)
        self.lons = lons; self.lats = lats; self.npts = npts
        self.x = x; self.y = y; self.z = z
        self.lptr = lptr; self.lst = lst; self.lend = lend
        self._shuffle=False; self._ix=None
    def interp(self,olons,olats,data,order=1):
        """
given a triangulation, perform interpolation on
output mesh defined by olons,olats (in radians), return result in data.
olons, olats can be 1d or 2d (output data array has same shape as olats,lons).
order of interpolation specified by 'order' kwarg, can be 0 (nearest neighbor),
1 (linear), or 3 (hermite cubic).

Algorithms:
 R. J. Renka, "ALGORITHM 623:  Interpolation on the Surface of a
 Sphere", ACM Trans. Math. Software, Vol. 10, No. 4, December 1984,
 pp. 437-439.
"""
        shapeout = olons.shape
        if len(shapeout) not in [1,2]:
            raise ValueError('olons,olats must be 1d or 2d')
        olons1 = (olons.astype(np.float64,copy=False)).ravel()
        olats1 = (olats.astype(np.float64,copy=False)).ravel()
        nptso = len(olons1)
        if (np.abs(olons1)).max() > 2.*np.pi:
            msg="lons must be in radians (-2*pi <= lon <= 2*pi)"
            raise ValueError(msg)
        if (np.abs(olats1)).max() > 0.5*np.pi:
            msg="lats must be in radians (-pi/2 <= lat <= pi/2)"
            raise ValueError(msg)
        if len(olats1) != nptso:
            raise ValueError('lons and lats must have same length')
        if len(data) != self.npts:
            raise ValueError('input data wrong size')
        if order not in [0,1,3]:
            raise ValueError('order must be 0,1 or 3')
        else:
            odata,ierr = \
            _stripack.interp_n(order, olats1, olons1,\
            self.x, self.y, self.z, data.astype(np.float64),\
            self.lst,self.lptr,self.lend,self.npts,nptso)
        if ierr != 0:
            raise ValueError('ierr = %s in intrpc0_n' % ierr)
        return odata.reshape(shapeout)
    def interp_nn(self,olons,olats,data):
        """
same as interp(olons,olats,data,order=0)"""
        return self.interp(olons,olats,data,order=0)
    def interp_linear(self,olons,olats,data):
        """
same as interp(olons,olats,data,order=1)"""
        return self.interp(olons,olats,data,order=1)
    def interp_cubic(self,olons,olats,data):
        """
same as interp(olons,olats,data,order=3)"""
        return self.interp(olons,olats,data,order=3)
    def interp_latlon(self,nlons,data,order=1,quiet=False):
        """
convenience method to generate a reg lat/lon output grid and do interpolation.

nlons:  number of longitudes on output grid
        (a reg lat/lon grid with nlons longitudes and
         nlons/2 latitudes, not including poles or wrap-
         around longitude).
data: 1d array of cubed sphere data
order: order of interpolation (0 for nearest neighbor, 
       1 for linear, 3 for cubic - default is 1).
quiet: Suppress diagnostic messages (default False)

returns lons2d,lats2d,latlon_data where
lons2d: 2d array of longitudes on output grid (in degrees)
lats2d: 2d array of latiitudes on output grid (in degrees)
latlon_data:  2d array of interpolated data on output grid.
        """
        # generate regular 2d lat/lon grid (not including poles,
        # or wrap-around longitude).
        nlats = nlons/2; delta = 360./nlons
        olons = delta*np.arange(nlons)
        olats = -90. + delta*(0.5 + np.arange(nlats))
        olonsd, olatsd = np.meshgrid(olons, olats) # degrees
        # interpolate to the reg lat/lon grid
        if not quiet: t1 = time.time()
        if self._shuffle:
            data = data[self._ix] # points were randomly shuffled to make triangulation faster
        latlon_data = self.interp(np.radians(olonsd),np.radians(olatsd),data,order=order) # expects radians
        if not quiet:
            t = time.time()-t1
            print('time to interpolate to %s by %s lat/lon grid = %s' %
                  (nlons,nlats,t))
            print('shape/min/max',latlon_data.shape, latlon_data.min(), latlon_data.max())
        return olonsd,olatsd,latlon_data

if __name__ == "__main__":
    # test function.
    def _test(npts=14000,nlons=360,tol=[0.2,1.e-2,None,3.e-3]):
        def fibonacci_pts(npts):
            # return lats and lons of N=npts fibonacci grid on a sphere.
            pi = np.pi
            inc = pi * (3.0 - np.sqrt(5.0))
            off = 2. / npts
            lats = []; lons = []
            for k in range(npts):
               y = k*off - 1. + 0.5*off
               r = np.sqrt(1 - y**2)
               phi = k * inc
               x = np.cos(phi)*r
               z = np.sin(phi)*r
               theta = np.arctan2(np.sqrt(x**2+y**2),z)
               phi = np.arctan2(y,x)
               lats.append( 0.5*pi-theta )
               if phi < 0.: phi = 2.*pi+phi
               lons.append( phi )
            return np.array(lats), np.array(lons)
        # fake test data.
        def test_func(lon, lat):
            nexp = 8
            return np.cos(nexp*lon)*np.sin(0.5*lon)**nexp*np.cos(lat)**nexp+np.sin(lat)**nexp
        # function to check error
        def check_err(latlon_data, latlon_datax, order):
            err = (np.abs(latlon_datax-latlon_data)).max()
            print('order = %s: max abs error %s specified tolerance %s' %\
                    (order,err,tol[order]))
            assert(err < tol[order])
        # input mesh (fibonacci spiral points)
        lats, lons = fibonacci_pts(npts)
        icos_data = test_func(lons,lats)
        # triangulation
        tri = trmesh(lons, lats)
        assert( tri.npts == npts )
        assert( np.array_equal(tri.lons, lons) )
        assert( np.array_equal(tri.lats, lats) )
        # output mesh
        nlats = nlons//2; delta = 360./nlons
        olons = delta*np.arange(nlons)
        olats = -90. + delta*(0.5 + np.arange(nlats))
        olons = np.radians(olons);  olats = np.radians(olats)
        olons, olats = np.meshgrid(olons, olats)
        # nearest neighbor interpolation
        order = 0 # can be 0 (nearest neighbor) or 1 (linear) or 3 (cubic)
        latlon_data = tri.interp(olons,olats,icos_data,order=order)
        # check error
        latlon_datax = test_func(olons,olats)
        check_err(latlon_data,test_func(olons,olats),order)
        # test interp_nn alias.
        latlon_data = tri.interp_nn(olons,olats,icos_data)
        latlon_datax = test_func(olons,olats)
        check_err(latlon_data,test_func(olons,olats),order)
        # linear interpolation
        order = 1 # can be 0 (nearest neighbor) or 1 (linear) or 3 (cubic)
        latlon_data = tri.interp(olons,olats,icos_data,order=order)
        # check error
        check_err(latlon_data,test_func(olons,olats),order)
        # test interp_linear alias
        latlon_data = tri.interp_linear(olons,olats,icos_data)
        latlon_datax = test_func(olons,olats)
        check_err(latlon_data,test_func(olons,olats),order)
        # cubic interpolation
        order = 3 # can be 0 (nearest neighbor) or 1 (linear) or 3 (cubic)
        latlon_data = tri.interp(olons,olats,icos_data,order=order)
        # check error
        latlon_datax = test_func(olons,olats)
        check_err(latlon_data,latlon_datax,order)
        # test interp_cubic alias
        latlon_data = tri.interp_cubic(olons,olats,icos_data)
        check_err(latlon_data,latlon_datax,order)
        # check convenience method
        lons2d, lats2d, latlon_data =\
        tri.interp_latlon(nlons,icos_data,quiet=True)
        check_err(latlon_data,latlon_datax,1)
    import unittest
    class RegridTest(unittest.TestCase):
        def test(self):
            _test()
    unittest.main()
